### PR TITLE
Emit condition check at the end of loop

### PR DESCRIPTION
**Pull request by kerams**
Fixes https://github.com/dotnet/fsharp/issues/12138.

The IL produced by F# and C# for the given snippet is now identical. Not sure what (if anything) this means for debug points.

Is there a way to automatically regenerate the offending baseline files?

### CODE REVIEW

**Comment by smoothdeveloper** [comment]
> Is there a way to automatically regenerate the offending baseline files?

Not sure if you tried this: https://github.com/dotnet/fsharp/blob/main/DEVGUIDE.md#updating-baselines-in-tests

If you are able to run those tests locally, it will create the "actual" files that you can then rename over the "baseline" ones (that will then show up in the diff).

**Comment by kerams** [comment]
Hmm, weird that some of the updated baselines still fail.

**Comment by dsyme** [comment]
@kerams Did you update using release mode, e.g.

    .\build -testFSharpQA -c Release

plus also the other test suites?  Thanks.  There can be a codegen difference

**Comment by dsyme** [comment]
We will need to check debugging experience manually.  

* For loops and while loops in synchronous code
* For loops and while loops in task code
* For loops and while loops in list expressions and array expressions
* For loops and while loops in sequence expressions

This is one way on Windows

    artifacts\bin\fsc\Debug\net472\fsc.exe --optimize- -tailcalls- tests\walkthroughs\DebugStepping\TheBigFileOfDebugStepping.fsx

    devenv /debugexe TheBigFileOfDebugStepping.exe

and then find the above constructs, breakpoint on the function defining them, and check F11 stepping behaviour is as before.

**Comment by kerams** [comment]
@dsyme, thanks. the debug point needed moving in front of the condition. While loop stepping works now but I'm not 100% sure about foreach, so another set of eyes would be appreciated.

Additionally, what's going on with [these tests](https://dev.azure.com/dnceng/public/_build/results?buildId=1707583&view=ms.vss-test-web.build-test-results-tab&runId=46480266&resultId=100819&paneView=debug)? The message is incredibly cryptic. The build did generate new `providedDesigner.dll` for me, but the tests still fail locally, so I did not commit the files.

**Comment by dsyme** [comment]
@kerams here; https://stackoverflow.com/questions/13274384/why-is-this-net-il-not-verifiable

I'd expect you're hitting this in situations where there is something already on the IL stack.  There could be a repro in code like this:

```fsharp
let f() =
    callSomething firstArgument (while .... do ...)
```

Here `firstArgument` will be pushed and then the `while` loop generated.  

It might be you have to either avoid doing the codegen when the IL stack is non-empty.  Or else you have to save the stack and restore it, which we already do for integer for-loops https://github.com/dotnet/fsharp/pull/12959/files#diff-bff3f7337de1c9a878112e73f774cc564bcae3e1e4785c8b1beb2af6cb8f6de9L4026

**Comment by dsyme** [comment]
> @dsyme, thanks. the debug point needed moving in front of the condition. While loop stepping works now but I'm not 100% sure about foreach, so another set of eyes would be appreciated.

@kerams Great. Let's make sure we have me or someone else do an end-to-end double check of the debug stepping in all the necessary cases

**Comment by kerams** [comment]
> It might be you have to either avoid doing the codegen when the IL stack is non-empty. Or else you have to save the stack and restore it, which we already do for integer for-loops

That was it :). I've decided to go the simpler route and fall back to using the old implementation when the stack is not empty. As you're saying, it's a bit of an edge case and I suspect saving/restoring the stack could well make the whole thing even slower than it is, which would defeat the purpose of the PR.

Ready for review, I think.

File: src/fsharp/IlxGen.fs
Comment (vzarytovskii) [inline src/fsharp/IlxGen.fs:4114]: Will it be a lot of work to have two separate tests in `FSharp.Compiler.ComponentTests\EmittedIL` which will verify both branches, to see if we have expected codegen?
Thanks.

Comment (kerams) [reply src/fsharp/IlxGen.fs:4114]: 2 such tests are in the new WhileLoops.fs. It's in ComponentTests, not as a baseline file though.

Comment (vzarytovskii) [reply src/fsharp/IlxGen.fs:4114]: > 2 such tests are in the new WhileLoops.fs

Oh, apologies, it seems that GH just collapsed the file and I missed it.
Thanks!

**Comment by dsyme** [comment]
Looks like this has unfortunately collided with moving some tests , could you fix things up @kerams ? Thanks

**Comment by kerams** [comment]
@dsyme, done.

**Review by vzarytovskii — APPROVED** [review]
This looks good to me, thanks @kerams!
Did you, by any chance, look at any JIT benchmarks, were there any changes?

**Comment by kerams** [comment]
I didn't, but since the IL is now aligned with C#, we should see similar results similar to what was reported in the original issue.

**Comment by dsyme** [comment]
@vzarytovskii Note we still need to double-check the debug experience here for the various cases I listed above https://github.com/dotnet/fsharp/pull/12959#issuecomment-1092388448

**Comment by vzarytovskii** [comment]
> @vzarytovskii Note we still need to double-check the debug experience here for the various cases I listed above [#12959 (comment)](https://github.com/dotnet/fsharp/pull/12959#issuecomment-1092388448)

Oh, I did go thru the `TheBigFileOfDebugStepping` before merging and verified `for` and `while` in synchronous code, task builder, list, array and seq expressions.

Sorry, forgot to mention it.

**Pull request merged**
